Source/Core/Core/GeckoCodeConfig.cpp

```diff
@@ -2,17 +2,173 @@
 // Licensed under GPLv2+
 // Refer to the license.txt file included.
 
+#include "Core/GeckoCodeConfig.h"
+
 #include <algorithm>
+#include <cstddef>
+#include <memory>
 #include <sstream>
-#include <string>
-#include <vector>
+
+#include <curl/curl.h>
 
 #include "Common/IniFile.h"
+#include "Common/Logging/Log.h"
 #include "Common/StringUtil.h"
-#include "Core/GeckoCodeConfig.h"
 
 namespace Gecko
 {
+static std::string MakeEndpointURL(const std::string& game_id)
+{
+  constexpr const char* url_prefix = "http://geckocodes.org/txt.php?txt=";
+
+  switch (game_id[0])
+  {
+  case 'R':
+  case 'S':
+  case 'G':
+    return url_prefix + game_id;
+  default:
+    // All channels (WiiWare, VirtualConsole, etc) are identified by their first four characters
+    return url_prefix + game_id.substr(0, 4);
+  }
+}
+
+static size_t DownloadCodesWriteCallback(void* contents, size_t size, size_t nmemb,
```

Comment:
Weird, the example they had on the site wrote the first parameter as `void*`.

Which is better: to have the last parameter here be `void*` and cast to `std::string*` inside the function; or cast the function to `size_t(*)(char*, size_t, size_t, void*)` when passing it to Curl?